qiskit/transpiler/passes/utils/merge_adjacent_barriers.py

```diff
@@ -69,29 +69,20 @@ def run(self, dag):
         if not node_to_barrier_qubits:
             return dag
 
-        # add the merged barriers to a new DAG
-        new_dag = dag.copy_empty_like()
-
-        # go over current nodes, and add them to the new dag
-        for node in dag.topological_op_nodes():
-            if node.name == "barrier":
-                if node in node_to_barrier_qubits:
-                    qubits = node_to_barrier_qubits[node]
-                    # qubits are stored as a set, need to convert to a list
-                    new_dag.apply_operation_back(
-                        Barrier(len(qubits)), qargs=sorted(qubits, key=indices.get), check=False
-                    )
-            else:
-                # copy the condition over too
-                new_dag.apply_operation_back(node.op, node.qargs, node.cargs, check=False)
-        return new_dag
+        for barrier in barriers:
+            if barrier in node_to_barrier_qubits:
+                barrier_to_add, nodes = node_to_barrier_qubits[barrier]
+                dag.replace_block_with_op(nodes, barrier_to_add, wire_pos_map=indices)
```

Comment:
I think we should be safe to set `cycle_check` to false here. The cycle checking adds runtime overhead here to validate the contraction doesn't introduce a cycle in the dag. But because we determined the overlap and adjacency in the collection I can't see a scenario where we'd end up with a cycle merging barriers.

```suggestion
                dag.replace_block_with_op(nodes, barrier_to_add, wire_pos_map=indices, cycle_check=False)
```